### PR TITLE
GH-2198: Spring Observability Initial Commit

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-kafka/issues/2198

### CODE REVIEW

**Comment by garyrussell** [comment]
Screen Shot 2022-09-15 at 10 03 22 AM

File: spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/DefaultKafkaListenerObservationConvention.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	@Override
+	public KeyValues getLowCardinalityKeyValues(KafkaRecordReceiverContext context) {
+		return KeyValues.of(KafkaListenerObservation.ListenerLowCardinalityTags.LISTENER_ID.asString(),
+						context.getListenerId());
+	}
+
+	@Override
+	public KeyValues getHighCardinalityKeyValues(KafkaRecordReceiverContext context) {
+		return KeyValues.empty();
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/DefaultKafkaListenerObservationConvention.java:44]: It looks like `default` method on the `ObservationConvention` does exactly the same

File: spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaTemplateObservation.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	public enum TemplateLowCardinalityTags implements KeyName {
+
+		/**
+		 * Bean name of the template.
+		 */
+		BEAN_NAME {
+
+			@Override
+			public String asString() {
+				return "bean.name";
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaTemplateObservation.java:68]: In Spring Integration PR @marcingrzejszczak has requested me to add a meaningful prefix to all the tags I expose on the observation: https://github.com/spring-projects/spring-integration/pull/3879#pullrequestreview-1104184375.
So, probably this one should go as a `spring.kafka.template.name` ?

File: spring-kafka/src/test/java/org/springframework/kafka/support/micrometer/ObservationTests.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+import io.micrometer.tracing.test.simple.SimpleSpan;
+import io.micrometer.tracing.test.simple.SimpleTracer;
+
+/**
+ * @author Gary Russell
+ * @since 3.0
+ *
+ */
+@SpringJUnitConfig
+@EmbeddedKafka(topics = { "observation.testT1", "observation.testT2" })
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/support/micrometer/ObservationTests.java:76]: `@DirtiesContext` to stop an embedded broker in the end?

**Review by artembilan — CHANGES_REQUESTED** [review]

**Review by artembilan — APPROVED** [review]

**Comment by garyrussell** [comment]
With async stop...

Screen Shot 2022-09-15 at 3 58 53 PM

**Review by artembilan — CHANGES_REQUESTED** [review]
This one still fails on the GH actions.
Probably because of that async `stop()` feature you try to incorporate.

@marcingrzejszczak,

would you mind to look into this PR as well?
Should be similar to what we have so far in Spring AMQP.

Thank you!

File: spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaRecordReceiverContext.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		this.record = record;
+		this.listenerId = listenerId;
+	}
+
+	public String getListenerId() {
+		return this.listenerId;
+	}
+
+	@Override
+	public String getName() {
```
Comment (marcingrzejszczak) [inline spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaRecordReceiverContext.java:57]: Don't set the name here - please set it via the convention

File: spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaRecordSenderContext.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		this.destination = record.topic();
+	}
+
+	public String getBeanName() {
+		return this.beanName;
+	}
+
+	@Override
+	public String getName() {
+		return "spring.kafka.template";
```
Comment (marcingrzejszczak) [inline spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaRecordSenderContext.java:51]: Don't set the name here - please set it via the convention

**Review by marcingrzejszczak — CHANGES_REQUESTED** [review]
I've found 2 minor issues, other than that LGTM.

The question is if you want to use the automated documentation generation - https://micrometer.io/docs/tracing#_documentation_building , that way all of your observations would be automatically scanned and an adoc file will be created cc @artembilan

**Pull request merged**